render/vulkan/vulkan.c

```diff
@@ -33,6 +33,81 @@ static const char *find_extensions(const VkExtensionProperties *avail,
 	return NULL;
 }
 
+static char **get_validation_layers(uint32_t *layer_count) {
+	// TODO request layer via env var
+	static const char *layers[] = {
+		"VK_LAYER_KHRONOS_validation",
+		// "VK_LAYER_RENDERDOC_Capture",
+		// "VK_LAYER_live_introspection",
+	};
+
+	static const size_t layers_len = sizeof(layers) / sizeof(layers[0]);
+
+	VkLayerProperties *layer_props = NULL;
+	char **found_layers = NULL;
+	uint32_t found_layers_count = 0;
+
+	uint32_t count;
+	vkEnumerateInstanceLayerProperties(&count, NULL);
+
+	if (count == 0) {
+		wlr_log(WLR_DEBUG, "No validation layers found");
+		goto layers_err;
+	}
+	wlr_log(WLR_DEBUG, "Found %"PRIu32" validation layers", count);
+
+	layer_props = calloc((size_t)count, sizeof(VkLayerProperties));
+	if (layer_props == NULL) {
+		wlr_log(WLR_ERROR, "Failed to allocate %"PRIu32" VkLayerProperties",
+			count);
+		goto layers_err;
+	}
+
+	found_layers = calloc((size_t)count, sizeof(char*));
+	if (found_layers == NULL) {
+		wlr_log(WLR_ERROR, "Failed to allocate validation layers");
+		goto layers_err;
+	}
+
+	vkEnumerateInstanceLayerProperties(&count, layer_props);
+	for (size_t i = 0; i < (size_t)count; ++i) {
+		wlr_log(WLR_DEBUG, "Vulkan instance validation layer %s v%"PRIu32,
+			layer_props[i].layerName, layer_props[i].implementationVersion);
+		for (size_t j = 0; j < layers_len; ++j) {
+			if (strcmp(layer_props[i].layerName, layers[j]) == 0) {
+				found_layers[found_layers_count] = calloc(
```

Comment:
The alloc and strcpy isn't needed. You know that `layer_props[i].layerName` is the same as `layers[j]` and the string in `layers[j]` is a string literal so you can just store that here. Then you can also remove the free calls for the layer names below.